text/0419-cdk-launchpads.md

```diff
@@ -0,0 +1,629 @@
+# CDK Launchpads
+
+## Working Backwards
+
+CDK Launchpads is a new functionality of the CDK project.
+It is geared primarily towards platform or system administrators in enterprise settings who are responsible for managing
+the AWS environments that are then used by application developers to deploy CDK applications into.
+
+Launchpads can be thought of as an extension of the
+[`cdk bootstrap` command](https://docs.aws.amazon.com/cdk/v2/guide/bootstrapping.html).
+The bootstrap command creates a set of resources -- S3 Bucket with a KMS Key, ECR Repository, several IAM Roles, etc. --
+that allow CDK applications to be deployed into a given AWS account and region.
+Launchpads are extensions of those bootstrap Stacks.
+They contain (by default) the same set of resources,
+but also add several features that are not available in the bootstrap command, such as:
+
+1. Ability to define compliance rules.
+  It is impossible to deploy a CDK application to the given environment if it breaks any compliance rule defined in the Launchpad.
+2. Easy customization of the bootstrap resources, like adding lifecycle policies to the ECR repository,
+  changing the names of the generated Roles, or any other customization needed.
+  With the existing bootstrap command,
+  the only way these customizations are supported is by providing your own CloudFormation template.
+3. Extending the predefined set of bootstrap resources with additional resources.
+  Common examples are VPC networks, IAM Roles or Permission Boundaries, or Security Groups.
+  These shared resources can then be easily referenced from CDK code by developers working in a given Launchpad
+  (see below for details).
+4. Launchpads integrate with AWS organizations,
+  allowing easy creation of new accounts that have the given Launchpad provisioned when the account is created,
+  alleviating the need for running any command, including `cdk bootstrap`, manually.
+5. Launchpads can be configured to use a given Managed Policy as a Permission Boundary for all CDK applications deployed into the given environment.
+
+Using Launchpads does not require any knowledge of the CDK --
+in fact, it does not require knowing a programming language at all!
+
+The functionality of CDK Launchpads can be accessed in three different ways:
+
+1. Through a dedicated CLI, `cdk-lpads`. This is a self-contained program
```

Comment:
> Is the only reason for the CLI that it can be installed without NodeJS? Maybe we should just rewrite the CDK CLI to not depend on NodeJS.

The idea is also to simplify (the CDK CLI has a _ton_ of stuff in it that's unrelated to Launchpads - removing that complexity seems valuable).

> Also, I'm wondering if this functionality needs to exist separate from `cdk bootstrap` at all?

It's possible we should make this functionality part of the `cdk bootstrap` command - I think that's a valid idea. The question is: is it a more ergonomic design, than a separate command? Historically, we were very hesitant to add more options to the `bootstrap` command, and that's why I went with a separate command in this proposal. But we can certainly discuss this - we can always change our mind on making more changes to `cdk bootstrap`.

> It almost feels like launchpads should exist as a group of construct libraries (or CFN public resources) where we don't necessarily provide an opinionated way to deploy them. What if you are an organization that already has a process for provisioning new accounts and deploying a baseline set of resources (maybe using something like [this](https://aws.amazon.com/solutions/implementations/customizations-for-aws-control-tower/))? You would want a way to provide a customizable CFN template that deploys CDK specific resources.

I think public CFN resources are out of the picture for now (too much development effort before we validate this idea), and CDK resources for this are planned (the RFC contains some examples), but we also need the ability to define Launchpads without the requirement of writing CDK code, to cater to a segment of customers who don't want to use a programming language.

Reply:
I personally like the idea of a simple command like bootstrap (or literally an extension of bootstrap) as the low-friction onboard, with a set of opinionated CDK constructs to allow for customization or a more comfortable IaC approach.

If the behavior of bootstrap is significantly different than launchpads, or would add a significant amount of overhead to bootstrap, I would endorse a seperate cmd. But it's largely a super set of bootstrap (which is sounds like it is), I'd love to have it be an opt in behavior of bootstrap.

Reply:
Here's a thought:

* If bootstrap resources were all in a construct...
* ...and therefore bootstrapping is equivalent to deploying a CDK app...
* ...then `cdk bootstrap` is the same as `cdk deploy`

And we don't need a special command at all anymore.

Reply:
Doesn't that lead to a "chicken and egg" problem though?

`cdk deploy` requires bootstrapping, but you can't bootstrap, because you can't `cdk deploy`?

Reply:
Are you sure that `cdk deploy` always needs bootstrapping? I thought without assets it works un-bootstrapped.

Reply:
Pretty sure that in V2, bootstrapping is required, even without Assets (there is a Rule that will make it fail if the version parameter stored in the bootstrap Stack is either not found, or not greater or equal than the expected version).